### PR TITLE
Changed enum values back to uint8 before updating the table

### LINKED ISSUE #55: Ensure adding to a non-defined table works
When working on https://github.com/PandABlocks/PandABlocks-ioc/pull/53 we wanted to ensure that all defined tables on the panda where used on `create_ioc`, with default values: `SEQ2:TABLE` will have its PVs even if the table received from the panda is `[]`.

When making the test for this became clear there were some problems with `_tables::update_table`... enums were required to be strings so the values could be pushed to the PVs with the current code, but softioc required floats. Trying to use the integer value for the enum would result in string float values for the PV: `["0.", "0.", "1."]`.

### PR DESCRIPTION
**Pull request by evalott100**
Closes #55 

Added a test to add to the panda table after initially leaving it blank

`update_table` will currently not work for enum values in the table since they're `List[str]` and have to be converted back to their integer index values.

Added a test to `update_table` for a table which is `[]` on ioc start up, then is added to on the panda side.

### CODE REVIEW

**Comment by evalott100** [comment]
Requires release of https://github.com/PandABlocks/PandABlocks-client/pull/58 before merging.

File: src/pandablocks_ioc/_tables.py [hunk 1]
```diff
@@ -679,7 +677,8 @@ def _update_scalar(self, waveform_record_name: str) -> None:
             scalar_val = waveform_data[index]
             if labels:
                 # mbbi/o records must use the numeric index
-                scalar_val = labels.index(scalar_val)
+                if isinstance(scalar_val, str):
+                    scalar_val = labels.index(scalar_val)
```
Comment (AlexanderWells-diamond) [inline src/pandablocks_ioc/_tables.py:681]: There should be a test of some kind to cover this line. Probably doesn't have to be a full system test.

File: tests/fixtures/panda_data.py [hunk 2]
```diff
@@ -212,8 +212,8 @@ def table_unpacked_data(
     """The unpacked equivalent of table_data"""
     array_values: List[ndarray] = [
         array([5, 0, 50000], dtype=uint16),
-        # Below labels correspond to numeric values [0, 6, 0]
-        array(["Immediate", "BITC=1", "Immediate"], dtype="<U9"),
+        # Below labels correspond to label values []"Immediate", "BITC=1", "Immediate"]
```
Comment (AlexanderWells-diamond) [inline tests/fixtures/panda_data.py:215]: Odd extra `]` here?

File: tests/test_ioc_system.py [hunk 3]
```diff
@@ -603,6 +603,34 @@ async def test_metadata_parses_into_multiple_pvs_caput_single_pv(
     ) in multiprocessing_queue_to_list(command_queue)
 
 
+async def test_non_defined_seq_table_can_be_added_to(
```
Comment (AlexanderWells-diamond) [inline tests/test_ioc_system.py:606]: I think it's probably worth writing a second system test, that has an empty SEQ3 and then tries to add data via writing PVs. Currently this only tests the flow in one direction.

**Review by AlexanderWells-diamond — CHANGES_REQUESTED** [review]
I think we need a couple more tests, especially as the one line of actual code change doesn't have a test covering it.

File: tests/test_ioc_system.py (hunk 3 above)
Comment (evalott100) [reply tests/test_ioc_system.py:606]: As discussed over slack, we'll handle this in a new PR.

**Comment by evalott100** [comment]
# TESTING

## On an old commit https://github.com/PandABlocks/PandABlocks-ioc/tree/a8586b9fde836a8a8eceda770c935f39af57f98c

### SEQ2 Table not defined yet
```
pandablocks-ioc softioc [test panda ip] PANDA outputted_screens

PANDA:SEQ2:TABLE:MODE          VIEW 
caput PANDA:SEQ2:TABLE:MODE 1
  Old : PANDA:SEQ2:TABLE:MODE          VIEW
  New : PANDA:SEQ2:TABLE:MODE          EDIT
caget PANDA:SEQ2:TABLE:TRIGGER
  PANDA:SEQ2:TABLE:TRIGGER 16384 0 0 0 0
caput -a PANDA:SEQ2:TABLE:TRIGGER 0 1 0
  Old : PANDA:SEQ2:TABLE:TRIGGER 16384 0 0 0
  New : PANDA:SEQ2:TABLE:TRIGGER 16384 1 0 0
caget PANDA:SEQ2:TABLE:TRIGGER
  PANDA:SEQ2:TABLE:TRIGGER 16384 1 0 0
caput PANDA:SEQ2:TABLE:MODE 2
  Old : PANDA:SEQ2:TABLE:MODE          EDIT
  New : PANDA:SEQ2:TABLE:MODE          SUBMIT
```
The ioc outputted
```
>>> WARNING:Value 0.0 of record SEQ2:TABLE:TRIGGER is not a recognised value.
Traceback (most recent call last):
  File "/scratch/twj43146/Programming/PandABlocks-ioc-fork/src/pandablocks_ioc/_tables.py", line 666, in _update_scalar
    scalar_val = labels.index(scalar_val)
ValueError: 0.0 is not in list
WARNING:Value 1.0 of record SEQ2:TABLE:TRIGGER is not a recognised value.
Traceback (most recent call last):
  File "/scratch/twj43146/Programming/PandABlocks-ioc-fork/src/pandablocks_ioc/_tables.py", line 666, in _update_scalar
    scalar_val = labels.index(scalar_val)
ValueError: 1.0 is not in list
INFO:Sending table data for SEQ2:TABLE to PandA
ERROR:Unable to Put record SEQ2:TABLE, value [],to PandA. Rolling back to last value from PandA.
Traceback (most recent call last):
  File "/scratch/twj43146/Programming/PandABlocks-ioc-fork/src/pandablocks_ioc/_tables.py", line 533, in update_mode
    packed_data = TablePacking.pack(
  File "/scratch/twj43146/Programming/PandABlocks-ioc-fork/src/pandablocks_ioc/_tables.py", line 171, in pack
    curr_val = [field_details.labels.index(x) for x in curr_val]
  File "/scratch/twj43146/Programming/PandABlocks-ioc-fork/src/pandablocks_ioc/_tables.py", line 171, in <listcomp>
    curr_val = [field_details.labels.index(x) for x in curr_val]
ValueError: 1.0 is not in list
```

### SEQ1 Table set up in web ui
```
pandablocks-ioc softioc [test panda ip] PANDA outputted_screens

PANDA:SEQ1:TABLE:MODE          VIEW 
caput PANDA:SEQ1:TABLE:MODE 1
  Old : PANDA:SEQ1:TABLE:MODE          VIEW
  New : PANDA:SEQ1:TABLE:MODE          EDIT
caget PANDA:SEQ1:TABLE:TRIGGER
  PANDA:SEQ1:TABLE:TRIGGER 16384 BITA=0 POSB>=POSITION BITA=1
caput -a PANDA:SEQ1:TABLE:TRIGGER 0 1 0
  Old : PANDA:SEQ1:TABLE:TRIGGER 16384 BITA=0 POSB>=POSITION BITA=1 
  New : PANDA:SEQ1:TABLE:TRIGGER 16384 1 0
caput PANDA:SEQ1:TABLE:MODE 2
Old : PANDA:SEQ1:TABLE:MODE          EDIT
New : PANDA:SEQ1:TABLE:MODE          SUBMIT
```

The ioc outputted
```
WARNING:Value 1 of record SEQ1:TABLE:TRIGGER is not a recognised value.
Traceback (most recent call last):
  File "/scratch/twj43146/Programming/PandABlocks-ioc-fork/src/pandablocks_ioc/_tables.py", line 666, in _update_scalar
    scalar_val = labels.index(scalar_val)
ValueError: '1' is not in list
INFO:Sending table data for SEQ1:TABLE to PandA
ERROR:Unable to Put record SEQ1:TABLE, value [],to PandA. Rolling back to last value from PandA.
Traceback (most recent call last):
  File "/scratch/twj43146/Programming/PandABlocks-ioc-fork/src/pandablocks_ioc/_tables.py", line 533, in update_mode
    packed_data = TablePacking.pack(
  File "/scratch/twj43146/Programming/PandABlocks-ioc-fork/src/pandablocks_ioc/_tables.py", line 171, in pack
    curr_val = [field_details.labels.index(x) for x in curr_val]
  File "/scratch/twj43146/Programming/PandABlocks-ioc-fork/src/pandablocks_ioc/_tables.py", line 171, in <listcomp>
    curr_val = [field_details.labels.index(x) for x in curr_val]
ValueError: '1' is not in list
```

None of the trigger values were changed on the webui.

## On this branch (`ensure_adding_to_a_non_defined_table_works`)

### SEQ2 Table not defined yet
```
pandablocks-ioc softioc 172.23.252.201 PANDA --screens-dir=outputted_screens --clear-bobfiles

PANDA:SEQ2:TABLE:MODE          VIEW 
caput PANDA:SEQ2:TABLE:MODE 1
  Old : PANDA:SEQ2:TABLE:MODE          VIEW
  New : PANDA:SEQ2:TABLE:MODE          EDIT
caget PANDA:SEQ2:TABLE:TRIGGER
  PANDA:SEQ2:TABLE:TRIGGER 16384 0 0 0 0
caput -a PANDA:SEQ2:TABLE:TRIGGER 0 1 0
  Old : PANDA:SEQ2:TABLE:TRIGGER 16384 0 0 0
  New : PANDA:SEQ2:TABLE:TRIGGER 16384 1 0 0
caget PANDA:SEQ2:TABLE:TRIGGER
  PANDA:SEQ2:TABLE:TRIGGER 16384 1 0 0
caput PANDA:SEQ2:TABLE:MODE 2
  Old : PANDA:SEQ2:TABLE:MODE          EDIT
  New : PANDA:SEQ2:TABLE:MODE          SUBMIT
```

The ioc outputted
```
>>> INFO:Sending table data for SEQ2:TABLE to PandA
ERROR:Unable to Put record SEQ2:TABLE, value [],to PandA. Rolling back to last value from PandA.
Traceback (most recent call last):
  File "/scratch/twj43146/Programming/PandABlocks-ioc-fork/src/pandablocks_ioc/_tables.py", line 555, in update_mode
    packed_data = table_to_words(table, self.field_info)
  File "/scratch/twj43146/Programming/PandABlocks-ioc-fork/.venv/lib/python3.10/site-packages/pandablocks/utils.py", line 111, in table_to_words
    assert len(packed) == len(column), (
AssertionError: Table record TRIGGER has mismatched length compared to other records, cannot pack data
```

### SEQ1 Table set up in web ui
```
PANDA:SEQ1:TABLE:MODE          VIEW
caput PANDA:SEQ1:TABLE:MODE 1
  Old : PANDA:SEQ1:TABLE:MODE          VIEW
  New : PANDA:SEQ1:TABLE:MODE          EDIT
caget PANDA:SEQ1:TABLE:TRIGGER
  PANDA:SEQ1:TABLE:TRIGGER 16384 1 9 2
caput -a PANDA:SEQ1:TABLE:TRIGGER 0 1 0
  Old : PANDA:SEQ1:TABLE:TRIGGER 16384 1 9 2 0
  New : PANDA:SEQ1:TABLE:TRIGGER 16384 1 0 0 0
caput PANDA:SEQ1:TABLE:MODE 2
  Old : PANDA:SEQ1:TABLE:MODE          EDIT
  New : PANDA:SEQ1:TABLE:MODE          SUBMIT
```

The ioc outputted
```
>>> INFO:Sending table data for SEQ1:TABLE to PandA
ERROR:Unable to Put record SEQ1:TABLE, value [],to PandA. Rolling back to last value from PandA.
Traceback (most recent call last):
  File "/scratch/twj43146/Programming/PandABlocks-ioc-fork/src/pandablocks_ioc/_tables.py", line 555, in update_mode
    packed_data = table_to_words(table, self.field_info)
  File "/scratch/twj43146/Programming/PandABlocks-ioc-fork/.venv/lib/python3.10/site-packages/pandablocks/utils.py", line 111, in table_to_words
    assert len(packed) == len(column), (
AssertionError: Table record TRIGGER has mismatched length compared to other records, cannot pack data
```

**Review by AlexanderWells-diamond — APPROVED** [review]
Thank you for testing those situations. To me it looks like this branch is now behaving as expected: Attempting to add a new row by only extending 1 of the waveforms triggers an error. 

Do you think this PR is a suitable place to add a test that properly adds a row via EPICS, i.e. does a `caput` to each and every column in a table?

Otherwise, please fix the CI linting errors and then two minor cleanup comments, then this is good to go.

**Comment by evalott100** [comment]
Putting it back to how it was before all this,

`caput -a PANDA:SEQ1:TABLE:TRIGGER 3 Immediate Immediate "BITA=0"` works.

I've made a [new branch](https://github.com/evalott100/PandABlocks-client/tree/restore_enum_behaviour) on the client which allows for packing of empty tables (it gives things default values if columns are missing).

**Comment by evalott100** [comment]
https://github.com/PandABlocks/PandABlocks-client/pull/59

**Comment by evalott100** [comment]
> Thank you for testing those situations. To me it looks like this branch is now behaving as expected: Attempting to add a new row by only extending 1 of the waveforms triggers an error.
> 
> Do you think this PR is a suitable place to add a test that properly adds a row via EPICS, i.e. does a `caput` to each and every column in a table?
> 
> Otherwise, please fix the CI linting errors and then two minor cleanup comments, then this is good to go.

@AlexanderWells-diamond  I've now put this back to how it was, before any of these changes. Whenever PVs are changed internally we push the enums as strings.

**Pull request merged**